app/Http/Middleware/UploadCheck.php

```diff
@@ -17,134 +19,169 @@
-            }
-            if($no_error)
-                return true;
-
-            Logs::error(__METHOD__, __LINE__, 'Album ownership mismatch!');
-            return false;
-        }
-
-        return null;
-    }
-
-
-    /**
-     * Check if the user is authorized to do anything to that picture
-     *
-     * @param Request $request
-     * @param int $id
-     * @return ResponseFactory|Response|mixed
-     */
-    public function photo_check(Request $request, int $id)
-    {
-        if ($request->has('photoID')) {
-            $photoID = $request['photoID'];
-            $num = Photo::where('id','=',$photoID)->where('owner_id','=',$id)->count();
-            if ($num == 0) {
-                Logs::error(__METHOD__, __LINE__, 'Could not find specified photo');
-                return false;
-            }
-            return true;
-        }
-
-        if ($request->has('photoIDs'))
-        {
-            $photoIDs = $request['photoIDs'];
-
-            $photos = Photo::whereIn('id',explode(',', $photoIDs))->get();
-            if ($photos == null)
-            {
-                Logs::error(__METHOD__, __LINE__, 'Could not find specified photos');
-                return false;
-            }
-            $no_error = true;
-            foreach ($photos as $photo_t) {
-                // either you own the picture or it is in an album you own
-                $no_error &= (($photo_t->owner_id == $id) || ($photo_t->album != null && $photo_t->album->owner_id == $id));
-            }
-            if($no_error)
-                return true;
-
-            Logs::error(__METHOD__, __LINE__, 'Photos ownership mismatch!');
-            return false;
-        }
-    }
+	/**
+	 * @var SessionFunctions
+	 */
+	private $sessionFunctions;
+
+
+	/**
+	 * @var AlbumFunctions
+	 */
+	private $albumFunctions;
+
+
+	public function __construct(SessionFunctions $sessionFunctions, AlbumFunctions $albumFunctions)
+	{
+		$this->sessionFunctions = $sessionFunctions;
+		$this->albumFunctions = $albumFunctions;
+	}
+
+
+
+	/**
+	 * Handle an incoming request.
+	 *
+	 * @param Request $request
+	 * @param Closure $next
+	 * @return mixed
+	 */
+	public function handle(Request $request, Closure $next)
+	{
+		// not logged!
+		if (!$this->sessionFunctions->is_logged_in()) {
+			return response('false');
+		}
+
+		// is admin
+		if ($this->sessionFunctions->is_admin()) {
+			return $next($request);
+		}
+
+		$user_id = Session::get('UserID');
+		$user = User::find($user_id);
+		if ($user == null)
+		{
+			return response('false');
+		}
+
+		// is not admin and does not have upload rights
+		if (!$user->upload) {
+			return response('false');
+		}
+
+		$ret = $this->album_check($request, $user_id);
+		if ($ret === false) {
+			return response('false');
+		}
+
+		$ret = $this->photo_check($request, $user_id);
+		if ($ret === false) {
+			return response('false');
+		}
+
+		// why do we want this ?
+		// I don't remember >_<
```

Comment:
It will only be used for the `Sharing::Delete` API call. It seems to check if the sharing(s) to be deleted concern the album(s) owned by the current user.

Reply:
Yep gotcha. Thanks for bracktracking that one.